### DISCUSSION TITLE
Component Design Philosophy (Svelte)

### DISCUSSION THREAD

**Discussion by newlegendmedia** [pocketbase/pocketbase#319] [Q&A] [answered]
I'm looking to create some Svelte ui components for a couple apps I have planned and I'm wondering about your approach to making more reusable long-term UI components. I see you have some base components and more specialized versions. 

Could you share a little about your component design philosophy and ideas for how to design for ease of use and reuse?

Thanks for any ideas you could provide, love what you've done here!

**Comment by ganigeorgiev** [comment, marked as answer]
_I'm not sure that I'll be able to help you, but before giving a brief overview on the Admin UI structure, keep in mind that the current approach I follow may not be suitable for everyone and there are definitely places where it could be done better, especially when you target a single framework like in your case._

I usually first group by the type of the thing and then if necessary by its context, because this comes with the lowest mental load, at least for me, when you have to decide where to put something. Or in other words - I put all actions inside `actions/` directory, all stores inside `stores/` directory, etc. 

I don't think I follow a strict design philosophy.
In `components/base/` I just have placed all generic components that are commonly used within the entire application (eg. select dropdown).
The other components are grouped by their page/route context, eg. users page related components are inside the `components/users/` directory, where the page/route component itself is prefixed with `Page` for easier visiual lookup (or when searching with `Ctrl + P` for example).

One thing that you may have noticed is that the styles (the `scss/` directory) are not tightly coupled to the Svelte components.
This is where a lot of frontend developers would usually disagree with my approach and probably most of them would prefer to have the styles inside the component so that they don't have to load some external css before importing it.
This is done this way because initially I wasn't going to use Svelte (or any other frontend framework that required a bundler), since I was planning to make the interface pluggable with external user-land code, so that users can customize it without requiring a build step, aka. just by adding some plain old HTML and reusing my css classes. But I've decided to keep it as simple as possible for now and leave the ui customizations for later to consider based on the users feedback.

And that is it in general.

**The only advice I can give you is to not overengineer your components**. Take the below with a grain of salt:
- If you are using TypeScript, try to keep the exported props type as simple as possible and if some of them are classes make sure that at least it is easy to construct the required object (eg. has builtin factories).
- Use components composition carefully and don't rely too much on `$$props` and `$$restProps`.
- Whenever possible use css classes instead of `style:prop={...}`.
- Try to keep your components self-contained as much as possible so that the users are not required to combine them with other several components just so that they can use them (I often see this in accordions, tabs and other similar components).

**Reply by newlegendmedia** [reply]
Superb answer! 

I did notice the styles are separate (not how Svelte commonly handles it), but it makes sense to me because I had a useful set of styles in LESS and have used them similarly so far with Svelte. I am starting to work out a CSS variables approach for configuring/theming and applying the vars in each component's style block.

Browsing thru your project, I found the folder structure, naming, and code formatting to be very clear and easy to absorb.

I am going to refer to this as I try to work out some new components. Thanks for sharing it and good luck with your project!
